### PR TITLE
#99 Add finished jobs in a separate tab

### PR DESCRIPTION
**Pull request by Firesphere**
Also added the messages to the job view, by using a separate class.
The classname is updated before write when it's complete.

Updated test to fetch the FinishedJob instead of the QueuedJobDescriptor.

### CODE REVIEW

**Comment by nyeholt** [comment]
Is there any reason it couldn't just use a different filter mechanism on the main model admin view? Whereby the 'default' state of the filtering would be to only display completed jobs, but allows someone to untick that?

Then update the QueuedJobDescriptor's can* methods to account for the Complete status.

**Comment by Firesphere** [comment]
My thought was, to move them to a separate Class, would be a cleaner way to handle it.

File: code/dataobjects/FinishedJob.php [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+    {
+        return true;
+    }
+
+    public function canEdit($member = null)
+    {
+        return false;
+    }
+
+    public function canView($member = null)
```
Comment (nyeholt) [inline code/dataobjects/FinishedJob.php]: Can this be restricted to admin users only?

File: code/services/AbstractQueuedJob.php [hunk 2]
```diff
@@ -136,7 +136,7 @@ public function jobFinished() {
 	 * Called when the job is determined to be 'complete'
 	 */
 	public function afterComplete() {
-
+        $this->ClassName = 'FinishedJobs';
```
Comment (nyeholt) [inline code/services/AbstractQueuedJob.php]: FinishedJob~~s~~

**Comment by nyeholt** [comment]
My concern though is that it's adding additional complexity in the post-complete phase that's not really necessary for the usecase required. If there are additional things that could make use of the separate object, that might make it a useful addition

File: code/dataobjects/FinishedJob.php (hunk 1 above)
Comment (Firesphere) [inline code/dataobjects/FinishedJob.php]: Sure, note added to my list of things-to-do :)

File: code/services/AbstractQueuedJob.php (hunk 2 above)
Comment (Firesphere) [inline code/services/AbstractQueuedJob.php]: Good spot!

**Comment by Firesphere** [comment]
I would think, having the classname changed, would also make a cleanup job, as per #101 easier. Plus, the added benefit is that the job runner only has to scan the QueuedJobs and thereby not having to run through possibly late finished jobs.
The latter was an issue I ran into, when having jobs running that take longer than expected, and "restart" the job, multiple times due to that. By changing it's classname, the first run finishing late, would change the classname to finished and by then, that issue didn't occur anymore for any subsequent automated crontasks trying to run it.

**Comment by dhensby** [comment]
> I would think, having the classname changed, would also make a cleanup job, as per #101 easier. Plus, the added benefit is that the job runner only has to scan the QueuedJobs and thereby not having to run through possibly late finished jobs.

I actually think it makes things harder. At the moment that PR would allow devs to dictate which jobs can be cleaned by status. Now it'd only be able to clean up ones that completed and changed the class correctly - or it'd need refactoring to clean up multiple object types.

**Comment by nyeholt** [comment]
> when having jobs running that take longer than expected, and "restart" the job, multiple times due to 
> that. By changing it's classname, the first run finishing late, would change the classname to finished and 
> by then, that issue didn't occur anymore for any subsequent automated crontasks trying to run it.

So the classname changed before the job is actually fully complete? Because that could be very problematic - either way, I don't really like changing the classname as part of the execution. I'd really prefer a filter option on the side - and if there's a real need to have them as a separate job, that that happens as a post-fact archiving step.

**Comment by robbieaverill** [comment]
It doesn't sound like there's an agreement from the maintainers that this feature fits the project, and since it's pretty stale now I'm going to close it. Feel free to reopen if you'd like to revisit it again! Thanks for contributing =)

**Pull request closed without merging**